pkg/controllers/externalsecret/externalsecret_controller.go

```diff
@@ -128,345 +167,552 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
-		secret.SetGroupVersionKind(gvks[0])
+	// set the binding reference to the secret
+	// https://github.com/external-secrets/external-secrets/pull/2263
+	es.Status.Binding = v1.LocalObjectReference{Name: secretName}
+
+	// mutate a copy of the existing secret with the mutation function
+	updatedSecret := existingSecret.DeepCopy()
+	if err := mutationFunc(updatedSecret); err != nil {
+		return fmt.Errorf(errMutate, updatedSecret.Name, err)
 	}
 
-	if equality.Semantic.DeepEqual(existing, secret) {
+	// if the secret does not need to be updated, return early
+	if equality.Semantic.DeepEqual(existingSecret, updatedSecret) {
 		return nil
 	}
-	// Cleaning up Managed fields manually as to keep patch coherence
-	secret.ObjectMeta.ManagedFields = nil
-	// we're not able to resolve conflicts so we force ownership
-	// see: https://kubernetes.io/docs/reference/using-api/server-side-apply/#using-server-side-apply-in-a-controller
-	if err := r.Client.Patch(ctx, secret, client.Apply, client.FieldOwner(fqdn), client.ForceOwnership); err != nil {
-		return fmt.Errorf(errPolicyMergePatch, secret.Name, err)
+
+	// if the existing secret is immutable, we can only update the object metadata
+	if existingSecret.Immutable != nil && *existingSecret.Immutable {
+		// check if the metadata was changed
+		metadataChanged := !equality.Semantic.DeepEqual(existingSecret.ObjectMeta, updatedSecret.ObjectMeta)
+
+		// check if the immutable data/type was changed
+		var dataChanged bool
+		if metadataChanged {
+			// update the `existingSecret` object with the metadata from `updatedSecret`
+			// this lets us compare the objects to see if the immutable data/type was changed
+			existingSecret.ObjectMeta = *updatedSecret.ObjectMeta.DeepCopy()
+			dataChanged = !equality.Semantic.DeepEqual(existingSecret, updatedSecret)
+		} else {
+			// we know there was some change in the secret (or we would have returned early)
+			// we know the metadata was NOT changed (metadataChanged == false)
+			// so, the only thing that could have changed is the immutable data/type fields
+			dataChanged = true
+		}
+
+		// because we use labels and annotations to keep track of the secret,
+		// we need to update the metadata, even if the data is immutable
+		if metadataChanged {
+			// NOTE: we are using the `existingSecret` object here, as we ONLY want to update the metadata,
+			//       and we previously copied the metadata from the `updatedSecret` object
+			if err := r.Update(ctx, existingSecret, client.FieldOwner(fqdn)); err != nil {
+				// if we get a conflict, we should return early to requeue immediately
+				// note, we don't wrap this error so we can handle it in the caller
+				if apierrors.IsConflict(err) {
+					return err
+				}
+				return fmt.Errorf(errUpdate, existingSecret.Name, err)
+			}
+		}
+
+		// if the immutable data was changed, we should return an error
+		if dataChanged {
+			return fmt.Errorf(errUpdate, existingSecret.Name, ErrSecretImmutable)
+		}
```

Comment:
```suggestion
		// check if the immutable data/type was changed
		if metadataChanged {
			// update the `existingSecret` object with the metadata from `updatedSecret`
			// this lets us compare the objects to see if the immutable data/type was changed
			existingSecret.ObjectMeta = *updatedSecret.ObjectMeta.DeepCopy()
			if err := r.Update(ctx, existingSecret, client.FieldOwner(fqdn)); err != nil {
				// if we get a conflict, we should return early to requeue immediately
				// note, we don't wrap this error so we can handle it in the caller
				if apierrors.IsConflict(err) {
					return err
				}
				return fmt.Errorf(errUpdate, existingSecret.Name, err)
			}
		}
		
		// if the immutable data was changed, we should return an error
		if equality.Semantic.DeepEqual(existingSecret, updatedSecret) {
			return fmt.Errorf(errUpdate, existingSecret.Name, ErrSecretImmutable)
		}
```

I think this is much easier to follow than three ifs and an else. WDYT?

Reply:
These two pieces of code are not equivalent, with your changes we will always be doing a full deep equal twice, as we already return-early outside the immutable section if there was no change at all.

However, we can remove the second `if metadataChanged {` statement, because we can just include the update in the first if statment.

Reply:
I have removed the second `metdataChanged` if in https://github.com/external-secrets/external-secrets/pull/4086/commits/d40d76960900a1be45e13cb3110b06bc6cb9e84f

Reply:
>These two pieces of code are not equivalent, with your changes we will always be doing a full deep equal twice, as we already return-early outside the immutable section if there was no change at all.

You're right. I didn't think about the full two deepequal checks. Thanks for the fix! 🙇